org-ql-view.el

```diff
@@ -1032,39 +1035,48 @@ current buffer.  Otherwise BUFFERS-FILES is returned unchanged."
+              (pcase (expand-files buffers-files)
+                ((pred (seq-set-equal-p (mapcar #'expand-file-name (org-agenda-files))))
+                 "org-agenda-files")
+                ((and (guard (file-exists-p org-directory))
+                      (pred (seq-set-equal-p (org-ql-search-directories-files
+                                              :directories (list org-directory)))))
+                 "org-directory")
+                (_ buffers-files)))
+             ((pred (equal (current-buffer)))
+              "buffer")
+             ((or 'org-agenda-files '(function org-agenda-files))
               "org-agenda-files")
-             ((and (guard (file-exists-p org-directory))
-                   (pred (seq-set-equal-p (org-ql-search-directories-files
-                                           :directories (list org-directory)))))
-              "org-directory")
-             (_ buffers-files)))
-          ((pred (equal (current-buffer)))
-           "buffer")
-          ((or 'org-agenda-files '(function org-agenda-files))
-           "org-agenda-files")
-          ((and (pred bufferp) (guard (buffer-file-name buffers-files)))
-           (buffer-file-name buffers-files))
-          ((pred bufferp)
-           (buffer-name buffers-files))
-          (_ buffers-files))
-      ;; All items needs to be strings to pick duplicates when used with the extend conterpart.
-      ;; So making sure the buffers are convered to file names
-      (if (stringp it)
-          it
-        (-map
-         (lambda (buffer-file)
-           (if (bufferp buffer-file)
-               (--if-let (buffer-file-name buffer-file)
-                   it
-                 (buffer-name buffer-file))
-             buffer-file))
-         it)))))
+             ((and (pred bufferp) (guard (buffer-file-name buffers-files)))
+              (buffer-file-name buffers-files))
+             ((pred bufferp)
+              (buffer-name buffers-files))
+             (_ buffers-files))))
+      ;; To filter duplicates with the extend counterpart of this function,
+      ;; this needs to be a string or a list of string.
+      ;; Hence, making sure the buffers are convered to file names or buffer names.
+      ;; Using file-names when it's a file-buffer to avoid duplicates resulting from
+      ;; the file-buffer and file name being entered.
+      (cl-typecase contracted-buffers-files
+        (function contracted-buffers-files)
+        (string contracted-buffers-files)
+        (list (--> contracted-buffers-files
+                -flatten -non-nil
+                (--map
+                 (pcase-exhaustive it
+                   ((pred stringp) it)
+                   ((pred bufferp) (or (buffer-file-name it)
+                                       (buffer-name buffer-file))))
+                 it)
+                -uniq))
+        (t (error (format "Value %s is not a string, a valid function or a list of buffer/strings" contracted-buffers-files)))))))
```

Comment:
This is still a very long form that's hard to follow.  Please try to rewrite it in a more functional way.